sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/options/PipelineOptionsReflector.java

```diff
@@ -0,0 +1,179 @@
+      Class<? extends PipelineOptions> optionsInterface) {
+    Iterable<Method> methods = ReflectHelpers.getClosureOfMethodsOnInterface(optionsInterface);
+    Multimap<String, Method> propsToGetters = getPropertyNamesToGetters(methods);
+
+    ImmutableSet.Builder<Property> setBuilder = ImmutableSet.builder();
+    for (Map.Entry<String, Method> propAndGetter : propsToGetters.entries()) {
+      String prop = propAndGetter.getKey();
+      Method getter = propAndGetter.getValue();
+      Class<?> declaringClass = getter.getDeclaringClass();
+
+      if (!PipelineOptions.class.isAssignableFrom(declaringClass)) {
+        continue;
+      }
+
+      if (declaringClass.isAnnotationPresent(Hidden.class)) {
+        continue;
+      }
+
+      setBuilder.add(Property.of((Class<? extends PipelineOptions>) declaringClass, prop, getter));
+    }
+
+    return setBuilder.build();
+  }
+
+  /**
+   * Retrieve the full set of pipeline option properties visible within the type hierarchy of the
+   * input interfaces. An option is "visible" if:
+   *
+   * <ul>
+   *   <li>The option is defined within the interface hierarchy closure of the input
+   *   {@link PipelineOptions}.</li>
+   *   <li>The defining interface is not marked {@link Hidden}.</li>
+   * </ul>
+   */
+  static Set<Property> collectVisibleProperties(
+      Collection<Class<? extends PipelineOptions>> optionsInterfaces) {
+    ImmutableSet.Builder<Property> setBulder = ImmutableSet.builder();
+    for (Class<? extends PipelineOptions> optionsInterface : optionsInterfaces) {
+      setBulder.addAll(collectVisibleProperties(optionsInterface));
+    }
+
+    return setBulder.build();
+  }
+
+  /**
+   * Extra properties and their respective getter methods from a series of {@link Method methods}.
+   */
+  static Multimap<String, Method> getPropertyNamesToGetters(Iterable<Method> methods) {
+    Multimap<String, Method> propertyNamesToGetters = HashMultimap.create();
+    for (Method method : methods) {
+      String methodName = method.getName();
+      if ((!methodName.startsWith("get")
+          && !methodName.startsWith("is"))
+          || method.getParameterTypes().length != 0
+          || method.getReturnType() == void.class) {
+        continue;
+      }
+      String propertyName = Introspector.decapitalize(
+          methodName.startsWith("is") ? methodName.substring(2) : methodName.substring(3));
+      propertyNamesToGetters.put(propertyName, method);
+    }
+    return propertyNamesToGetters;
+  }
+
+  /**
+   * A property defined in a {@link PipelineOptions} interface.
+   */
+  static class Property {
+    private final Class<? extends PipelineOptions> clazz;
+    private final String name;
+    private final Method getter;
+
+    static Property of(Class<? extends PipelineOptions> clazz, String name, Method getter) {
+      return new Property(clazz, name, getter);
+    }
+
+    private Property(Class<? extends PipelineOptions> clazz, String name, Method getter) {
```

Comment:
This seems like a special construct for display data since a property is defined by the union of all interfaces it appears on. Since it is both a getter and a setter.